# R2Dtool

Genomics utilities for handling, integrating, and viualising isoform-mapped RNA feature data.    


  - **Integrate** transcriptome-mapped data: R2Dtool performs liftover of transcriptome-mapped RNA features to their corresponding genomic coordinates

  - **Annotate** transcriptome-mapped sites: R2Dtool annotates transcript-specific metatranscript coordinates and absolute and relative distances to annotated transcript landmarks, in an isoform-specific manner.

  - **Visualise** isoform-aware RNA feature distributions: R2Dtool introduces isoform-aware **metatranscript** plots and **metajunction** plots to study the positonal distribution of RNA features around annotated RNA landmarks.


> [!NOTE]
> [Link to R2Dtool preprint on biorXiv](https://doi.org/10.1101/2022.09.23.509222)    
> Sethi, A. J., Mateos, P. A., Hayashi, R., Shirokikh, N. & Eyras, E. R2Dtool: Positional Interpretation of RNA-Centric Information in the Context of Transcriptomic and Genomic Features. biorXiv 2022.09.23.509222 (2022)
> doi:10.1101/2022.09.23.509222.

# Contents 

   - [Usage](#usage)
       - [Handling isoform-mapped RNA sites](#handling-isoform-mapped-rna-sites)
       - [Visualising isoform-aware RNA feature metadistributions](#visualising-isoform-aware-rna-feature-metadistributions)
   - [Installation](#installation)
   - [Input and output data structure](#input-and-output-data-structure)
   - [Utilities: Convert CHEUI model II output to a BED-like input](#utilities)
   - [General notes](#general-notes)

     
# Usage

## Handling isoform-mapped RNA sites 

**Liftover** transcriptome-mapped features to genomic coordinates:

```
Usage: r2d liftover -i <input> -g <gtf>

Arguments:
    -i, --input <input>: Path to tab-separated transcriptome sites in BED format. 
    -g, --gff <annotation>: Path to gene structure annotation in GFF or GTF (requires -f gtf) format.

Options:
    -f, --format <FORMAT>: Structure of gene annotation used, either GFF or GTF (default: GFF).
    -H, --header: Indicates the input file has a header, which will be preserved in the output (default: false).
    -o, --output <OUTPUT>: Path to output file. If not specified, results will be printed to STDOUT.

```
- Liftover prepends 6 columns to the input file, containing the genome coordinates of the transcript features in BED format
- All data in the original input are preserved in the output and shifted by 6 columns 

**Annotate** transcriptome-mapped sites with isoform-specific distances to transcript landmarks:

```
Usage: r2d annotate -i <input> -g <gtf>

Arguments:
    -i, --input <input>: Path to tab-separated transcriptome sites in BED format. 
    -g, --gff <annotation>: Path to gene structure annotation in GFF or GTF (requires -f gtf) format.

Options:
    -f, --format <FORMAT>: Structure of gene annotation used, either GFF or GTF (default: GFF).
    -H, --header: Indicates the input file has a header, which will be preserved in the output (default: false).
    -o, --output <OUTPUT>: Path to output file. If not specified, results will be printed to STDOUT.

```

Annotation adds the following information to the epitranscriptomic sites as additional coluumns, relying on the gene structure GTF to generate these data.

```
transcript_biotype | gene_name | gene_id | tx_len | cds_len | utr5_len | utr3_len | cds_start | cds_end | tx_end  | rel_pos | abs_cds_start | abs_cds_end | up_junc_dist | down_junc_dist
```

- ```tx_len```, ```cds-len```, ```utr5-len``` and ```utr3-len```represents the lengths of the entire transcript, the coding sequence, and the 5' and 3' untranslated regions (UTRs)
- ```cds_start``` and ```cds_end``` represent the positions of the coding sequence start and end compared to the transcript.
- ```rel_pos``` represents the scaled metatrascript position of the given RNA feature, between 0 and 3, where 0 represents transcript start-site, 1 represents CDS start, 2 represent CDS end, and 3 represents the 3' transcript end. 
- ```abs_cds_start``` and ```abs_cds_end``` represent the absolute distance (in nt) of a given feature from the cds start and end
- ```up_junc_dist``` and ```down_junc_dist``` repreesnt the absolute distance (in nt) of a given site from the nearest upstream and downstream splice-junction contained in a given transcript

> [!NOTE]
> - ```liftover``` and ```annotate``` requires columns 1-6 to contain feature coordinates against a transcriptome reference
> - ```annotate``` can be perfomed before, but __not__ after, ```liftover```
> - The GTF or GFF3 file used must exactly match the transcriptome to which features are mapped 

## Visualising isoform-aware RNA feature metadistributions

**Plot** the **metatranscript** distribution of RNA features:

```
Rscript R2_plotMetaTranscript.R <sites> <output>  <filter field> <cutoff> <lower/upper>

positional arguments:
- path to annotated file generated by R2Dtool annotate 
- path to output plot (including extension, e.g. .svg or .png)
- filter field; column used to select significant sites
- cutoff; integer value to use for determining significant
- "lower"/"upper": Select sites that have a significanace _lower_ or _higher_ than the cutoff value in the significance field 

```

**Plot** the **metajunction** distribution of RNA features:

```
Rscript R2_plotMetaJunction.R <sites> <output>  <filter field> <cutoff> <lower/upper>

positional arguments:
- path to annotated file generated by R2Dtool annotate 
- path to output plot (including extension, e.g. .svg or .png)
- filter field; column used to select significant sites
- cutoff; integer value to use for determining significant
- "lower"/"upper": Select sites that have a significnace _lower_ or _higher_ than the cutoff value in the significance field 

```
# Installation and dependencies 

#### Compiling R2Dtool from source: 

```
# requires rust 
git clone git@github.com:comprna/R2Dtool.git && cd R2Dtool
cargo build --release

# add R2DTOOL to PATH
export PATH="$PATH:$(pwd)/target/release"
```

#### Testing R2Dtool installation 

```
# download and compile R2Dtool 
git clone git@github.com:comprna/R2Dtool.git && cd R2Dtool
cargo build --release
export PATH="$PATH:$(pwd)/target/release"

# annotate bed-like transcriptomic sites with metatranscript coordinates, distance to splice junctions, transcript structure and transcript biotype
r2d annotate ./test/out_CHEUI_modelII.bed ./test/GRCm39_subset.gtf ./test/out_CHEUI_modelII_annotated.bed

# liftover annotated transcriptomic sites to genomic coordinates
r2d lift ./test/out_CHEUI_modelII_annotated.bed ./test/GRCm39_subset.gtf ./test/out_CMII_annotated_lifted.bed
```

> Test data was generated using [cheui](https://github.com/comprna/CHEUI) and converted to bed-like coordinates using [R2Dtool utilities](https://github.com/comprna/R2Dtool/blob/main/scripts/cheui_to_bed.sh).

#### Dependencies 

> [!NOTE]
> R2Dtool's core ```liftover``` and ```annotation``` functionality are built into an executable program built on Rust, and dependencies are managed during compilation by Cargo. 
> R2Dtool's ```metatranscript``` and ```metajunction``` plotting functinality are implemented as executable R scripts, and depend on the following packages:

```
r==4.2.0
tidyverse==1.3.1
genomicFeatures==1.48.0
rtracklayer==1.56.0
```

# Input and output data structure

R2Dtool is designed to work with tab-delimited, plain text BED-like files (optionally) with a column names in row 1. Following UCSC's recommendations, input and output BED files should be tab-delimited plain-text:

- column 1 should represent transcript,
- column 2 and 3 should represent the coordinate of interest in zero-based, half open coordinates (e.g. the third nucleotide of a transcript is represented as (2,3),
- column 5, representing strand, is assumed as being (+)
- columns 4, and any column greater than 5 are preserved during annotation or liftover, and may be used to hold additional information, e.g. predictions of RNA modification stoichiometry, probability, etc.

The GTF annotation provided must contain identical gene structures used to generate the transcriptome, including identical transcript names in the FASTA header. One option is to use genomes, transcriptomes and gene structures from the same genome release. Another option is for users to generate their own transcriptome using a genome and gene structure file, e.g. using gffread.  

# Utilities

### Convert CHEUI model II output to a BED-like input

- This script transposes CHEUI coordinates by +3 (BED interval start) and +4 (BED interval end) to represent a single nucleotides, and rearranges columns to become bed-like.

```
bash cheui_to_bed.sh [cheui model II output file] [cheui_to_bed output file]
```

# General notes
- Transposes transcriptomic coordinates to their underlying genomic coordinates to enable the comparison of epitranscriptomic sites between overlapping transcript isoforms, and also to enable visualization of epitranscriptomic sites on a genome browser.
- Transcriptome to genome liftover relies only on transcript_id, site coordinates, and strand (columns 1:3,6)
- The annotation used must be in GTF format and correspond correctly to the transcriptome used
- Other columns aren't considered during annotation or liftover, and can be used to store additional information of interest
- A header containing column names is expected for annotation and liftover
- Issues with transcript_id and transcript_version? Try commenting out lines 32:33 and 40:41 in lift.R
- Using a GENCODE annotation? Try commenting out line 25 in lift.R and line 27 in annotate.R
- GENCODE annotations use 'transcript_type' whereas Ensembl annotations use 'transcript_biotype'?
- Because the output files will have a header, this may need to be removed prior to opening the file on some genome browsers.






```
                                          ┌──────── liftover ───────►   Genome-centric transcriptomic sites
                                          │
Isoform-centric transcriptomic sites  ────┼──────── annotate ───────►   Annotated isoform-centric transcriptomic sites
                                          │
       .=.                                └───── lift + annotate ───►   Annotated genome-centric transcriptomic sites
      '==c|
      [)-+|    <(RNA to DNA)
      //'_|        
 snd /]==;\                                                                                                                                                                     
```









