// use crate::parse_annotation::{Exon, read_gtf_file, read_gff_file};

// inputs: table of transcriptomic sites and valid GTF/GFF3 annotation
// outputs: original table, with additional columns corresponding to annotation.

// Add this new function for the annotate functionality
pub fn run_annotate(matches: &clap::ArgMatches) {
    println!("Running the annotate functionality...");
    // We will fill this function out together
}
